hw/arm/xnu.c

```diff
@@ -150,17 +152,127 @@ static void macho_dtb_node_process(DTBNode *node)
 
 }
 
-DTBNode* load_dtb_from_file(char *filename){
+// Extracts the payload from an im4p file. If the file is not an im4p file,
+// the raw file contents are returned. Exits if an error occurs.
+// See https://www.theiphonewiki.com/wiki/IMG4_File_Format for an overview
+// of the file format.
+static void extract_im4p_payload(const char* filename, const char* payload_type, uint8_t **data, uint32_t* length) {
     uint8_t *file_data = NULL;
     unsigned long fsize;
 
-    if (g_file_get_contents(filename, (char **)&file_data, &fsize, NULL)) {
-        DTBNode *root = load_dtb(file_data);
-        return root;
-    } else {
-        abort();
+    char errorDescription[ASN1_MAX_ERROR_DESCRIPTION_SIZE];
+    asn1_node img4_definitions = ASN1_TYPE_EMPTY;
+    asn1_node img4;
+    int ret;
+
+    if (!g_file_get_contents(filename, (char **)&file_data, &fsize, NULL)) {
+        error_report("Could not load %s data from file '%s'", payload_type, filename);
+        exit(EXIT_FAILURE);
+    }
+
+    if (asn1_array2tree(img4_definitions_array, &img4_definitions, errorDescription)) {
+        error_report("Could not initialize the ASN.1 parser: %s.", errorDescription);
+        exit(EXIT_FAILURE);
+    }
+
+    if ((ret = asn1_create_element(img4_definitions, "Img4.Img4Payload", &img4) != ASN1_SUCCESS)) {
+        error_report("Could not create an Img4Payload element: %d", ret);
+        exit(EXIT_FAILURE);
     }
+
+    if ((ret = asn1_der_decoding(&img4, (const uint8_t*)file_data, (uint32_t)fsize, errorDescription)) == ASN1_SUCCESS) {
+        char magic[4];
+        char type[4];
+        char description[128];
+        int len;
+
+        len = 4;
+        if ((ret = asn1_read_value(img4, "magic", magic, &len)) != ASN1_SUCCESS) {
+            error_report("Failed to read the im4p magic in file '%s': %d.", filename, ret);
+            exit(EXIT_FAILURE);
+        }
+
+        if (strncmp(magic, "IM4P", 4) != 0) {
+            error_report("Could parse ASN.1 data in file '%s' because it does not start with the IM4P header.", filename);
+            exit(EXIT_FAILURE);
+        }
+
+        len = 4;
+        if ((ret = asn1_read_value(img4, "type", type, &len)) != ASN1_SUCCESS) {
+            error_report("Failed to read the im4p type in file '%s': %d.", filename, ret);
+            exit(EXIT_FAILURE);
+        }
+
+        if (strncmp(type, payload_type, 4) != 0) {
+            error_report("Could parse ASN.1 data in file '%s' because it is not a %s object.", filename, payload_type);
```

Comment:
```suggestion
            error_report("Could parse ASN.1 data in file '%s' because it is not a %s object, expected %s object.", filename, payload_type, type);
```